src/libmongoc/src/mongoc/mongoc-shared.c

```diff
@@ -0,0 +1,166 @@
+{
+   bson_mutex_init (&g_shared_ptr_mtx);
+}
+
+static void
+_shared_ptr_lock ()
+{
+   bson_mutex_lock (&g_shared_ptr_mtx);
+}
+
+static void
+_shared_ptr_unlock ()
+{
+   bson_mutex_unlock (&g_shared_ptr_mtx);
+}
+
+void
+mongoc_shared_ptr_reset (mongoc_shared_ptr *const ptr,
+                         void *const pointee,
+                         void (*const dtor) (void *))
+{
+   BSON_ASSERT (ptr && "NULL given to mongoc_shared_ptr_reset()");
+   if (!mongoc_shared_ptr_is_null (*ptr)) {
+      /* Release the old value of the pointer, possibly destroying it */
+      mongoc_shared_ptr_reset_null (ptr);
+   }
+   ptr->ptr = pointee;
+   ptr->_aux = NULL;
+   /* Take the new value */
+   if (pointee != NULL) {
+      BSON_ASSERT (dtor != NULL);
+      ptr->_aux = bson_malloc0 (sizeof (_mongoc_shared_ptr_aux));
+      ptr->_aux->dtor = dtor;
+      ptr->_aux->refcount = 1;
+      ptr->_aux->managed = pointee;
+   }
+   bson_once (&g_shared_ptr_mtx_init_once, _init_mtx);
+}
+
+void
+mongoc_shared_ptr_assign (mongoc_shared_ptr *const out,
+                          mongoc_shared_ptr const from)
+{
+   BSON_ASSERT (out &&
+                "NULL given as output argument to mongoc_shared_ptr_assign()");
+   mongoc_shared_ptr_reset_null (out);
+   *out = mongoc_shared_ptr_copy (from);
+}
+
+mongoc_shared_ptr
+mongoc_shared_ptr_create (void *pointee, void (*destroy) (void *))
+{
+   mongoc_shared_ptr ret = MONGOC_SHARED_PTR_NULL;
+   mongoc_shared_ptr_reset (&ret, pointee, destroy);
+   return ret;
+}
+
+void
+mongoc_atomic_shared_ptr_store (mongoc_shared_ptr *const out,
+                                mongoc_shared_ptr const from)
+{
+   mongoc_shared_ptr prev = MONGOC_SHARED_PTR_NULL;
+   BSON_ASSERT (
+      out && "NULL given as output argument to mongoc_atomic_shared_ptr_store");
+
+   /* We are effectively "copying" the 'from' */
+   (void) mongoc_shared_ptr_copy (from);
+
+   _shared_ptr_lock ();
+   /* Do the exchange. Quick! */
+   prev = *out;
+   *out = from;
+   _shared_ptr_unlock ();
+
+   /* Free the pointer that we just overwrote */
+   mongoc_shared_ptr_reset_null (&prev);
+}
+
+mongoc_shared_ptr
+mongoc_atomic_shared_ptr_load (mongoc_shared_ptr const *ptr)
+{
+   mongoc_shared_ptr r;
+   BSON_ASSERT (ptr && "NULL given to _mongoc_shared_ptr_take_atomic()");
+   _shared_ptr_lock ();
+   r = mongoc_shared_ptr_copy (*ptr);
+   _shared_ptr_unlock ();
+   return r;
+}
+
+mongoc_shared_ptr
+mongoc_shared_ptr_copy (mongoc_shared_ptr const ptr)
+{
+   mongoc_shared_ptr ret = ptr;
+   if (!mongoc_shared_ptr_is_null (ptr)) {
+      bson_atomic_int_fetch_add (
+         &ret._aux->refcount, 1, bson_memory_order_relaxed);
+   }
+   return ret;
+}
+
+void
+mongoc_shared_ptr_reset_null (mongoc_shared_ptr *const ptr)
+{
+   BSON_ASSERT (ptr && "NULL given to mongoc_shared_ptr_reset_null()");
+   BSON_ASSERT (
+      !mongoc_shared_ptr_is_null (*ptr) &&
+      "Unbound mongoc_shared_ptr given to mongoc_shared_ptr_reset_null");
```

Comment:
```suggestion
   BSON_ASSERT (!mongoc_shared_ptr_is_null (*ptr));
```
I am unable to find precedent for including the reason for assertion within the conditional expression in the C driver. I am aware this is a pattern used to expose said message in the assertion message of most compilers. Is this a practice we are comfortable introducing into the codebase?

CC: @kevinAlbs

Reply:
This is mostly just a stylistic choice. I try to make assertions as loud and explanatory as possible, so it is immediately identifiable why a particular assertion was put in place, and why it might have fired.